File: src/main/java/com/verlumen/tradestream/strategies/doubleemacrossover/DoubleEmaCrossoverParamConfig.java
package com.verlumen.tradestream.strategies.doubleemacrossover;

import com.google.common.collect.ImmutableList;
import com.google.protobuf.Any;
import com.verlumen.tradestream.discovery.ChromosomeSpec;
import com.verlumen.tradestream.discovery.ParamConfig;
import com.verlumen.tradestream.strategies.DoubleEmaCrossoverParameters;
import com.verlumen.tradestream.strategies.StrategyType;
import io.jenetics.IntegerChromosome;
import io.jenetics.NumericChromosome;

public final class DoubleEmaCrossoverParamConfig implements ParamConfig {
  private static final ImmutableList<ChromosomeSpec<?>> SPECS =
      ImmutableList.of(
          ChromosomeSpec.ofInteger(2, 30), // Short EMA Period
          ChromosomeSpec.ofInteger(10, 100) // Long EMA Period
          );

  @Override
  public ImmutableList<ChromosomeSpec<?>> getChromosomeSpecs() {
    return SPECS;
  }

  @Override
  public Any createParameters(ImmutableList<? extends NumericChromosome<?, ?>> chromosomes) {
    if (chromosomes.size() != SPECS.size()) {
      throw new IllegalArgumentException(
          "Expected " + SPECS.size() + " chromosomes but got " + chromosomes.size());
    }

    IntegerChromosome shortEmaPeriodChrom = (IntegerChromosome) chromosomes.get(0);
    IntegerChromosome longEmaPeriodChrom = (IntegerChromosome) chromosomes.get(1);

    DoubleEmaCrossoverParameters parameters =
        DoubleEmaCrossoverParameters.newBuilder()
            .setShortEmaPeriod(shortEmaPeriodChrom.gene().allele())
            .setLongEmaPeriod(longEmaPeriodChrom.gene().allele())
            .build();

    return Any.pack(parameters);
  }

  @Override
  public ImmutableList<? extends NumericChromosome<?, ?>> initialChromosomes() {
    return SPECS.stream()
        .map(ChromosomeSpec::createChromosome)
        .collect(ImmutableList.toImmutableList());
  }
}

File: src/main/java/com/verlumen/tradestream/strategies/doubleemacrossover/DoubleEmaCrossoverParamConfig.kt
package com.verlumen.tradestream.strategies.doubleemacrossover

import com.google.common.collect.ImmutableList
import com.google.protobuf.Any
import com.verlumen.tradestream.discovery.ChromosomeSpec
import com.verlumen.tradestream.discovery.ParamConfig
import com.verlumen.tradestream.strategies.DoubleEmaCrossoverParameters
import io.jenetics.IntegerChromosome
import io.jenetics.NumericChromosome

class DoubleEmaCrossoverParamConfig : ParamConfig {
    companion object {
        private val SPECS = ImmutableList.of(
            ChromosomeSpec.ofInteger(2, 30), // Short EMA Period
            ChromosomeSpec.ofInteger(10, 100) // Long EMA Period
        )
    }

    override fun getChromosomeSpecs(): ImmutableList<ChromosomeSpec<*>> = SPECS

    override fun createParameters(chromosomes: ImmutableList<out NumericChromosome<*, *>>): Any {
        require(chromosomes.size == SPECS.size) {
            "Expected ${SPECS.size} chromosomes but got ${chromosomes.size}"
        }

        val shortEmaPeriodChrom = chromosomes[0] as IntegerChromosome
        val longEmaPeriodChrom = chromosomes[1] as IntegerChromosome

        val parameters = DoubleEmaCrossoverParameters.newBuilder()
            .setShortEmaPeriod(shortEmaPeriodChrom.gene().allele())
            .setLongEmaPeriod(longEmaPeriodChrom.gene().allele())
            .build()

        return Any.pack(parameters)
    }

    override fun initialChromosomes(): ImmutableList<out NumericChromosome<*, *>> {
        return SPECS.stream()
            .map { it.createChromosome() }
            .collect(ImmutableList.toImmutableList())
    }
}
